src/sync/mpsc.rs

```diff
@@ -0,0 +1,134 @@
+        Sender {
+            tx: self.tx.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+/// Shared state across the `Sender`(s) and `Receiver`.
+struct Inner {
+    /// Stores the `Receiver` task. Only one receiver task is used.
+    rx_task: Lock<Option<Task>>,
+    /// Count of senders, needed for determining when the last sender has been dropped.
+    sender_count: AtomicUsize,
+}
+
+/// Receiver for the channel.
+#[must_use = "streams do nothing unless polled"]
+pub struct Receiver<T> {
+    /// Receiver
+    rx: mpsc::Receiver<T>,
+    /// Inner
+    inner: Arc<Inner>,
+}
+
+impl<T> Sender<T> {
+    /// Send a message on the channel
+    pub fn send(&self, t: T) -> Result<(), mpsc::SendError<T>> {
+        match self.tx.send(t) {
+            Ok(_) => {
+                // if an rx_task exists, try to unpark
+                // otherwise someone else is unparking the Receiver
+                if let Some(mut opt_task) = self.inner.rx_task.try_lock() {
+                    if let Some(task) = opt_task.take() {
+                        task.unpark();
+                    }
+                }
+                Ok(())
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        // Each cloned sender calls `Drop` so we need to ensure that only the last sender being
+        // dropped will unpark the `Receiver`.
+        // The result is discarded because it could be updated between this line and the next.
+        let _ = self.inner.sender_count.fetch_sub(1, SeqCst);
+        // if it was one before, it *should* be zero now, but let's check just to make sure
+        if self.inner.sender_count.compare_exchange(0, 0, SeqCst, SeqCst).is_ok() {
+            // wake up receiver to go to disconnection event or get more data if there is any
+            if let Some(mut opt_task) = self.inner.rx_task.try_lock() {
+                if let Some(task) = opt_task.take() {
+                    task.unpark();
+                }
+            }
+        }
+    }
+}
+
+/// Creates a new multiple-producer, single-consumer channel. `Sender` can be cloned.
+///
+/// This channel does not implement back pressure.
+pub fn channel<T>() -> (Sender<T>, Receiver<T>) {
+    let (tx, rx) = mpsc::channel();
+    let inner = Arc::new(Inner {
+        rx_task: Lock::new(None),
+        sender_count: AtomicUsize::new(1),
+    });
+    let tx = Sender {
+        tx: tx,
+        inner: inner.clone(),
+    };
+    let rx = Receiver {
+        rx: rx,
+        inner: inner,
+    };
+    (tx, rx)
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+    type Error = ();
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.rx.try_recv() {
+            Ok(t) => Ok(Some(t).into()),
+            Err(e) => {
+                match e {
+                    mpsc::TryRecvError::Disconnected => Ok(Async::Ready(None)),
```

Comment:
This may look a bit nicer as a nested match (e.g. match `Err(TryRecvError::...)`) above.